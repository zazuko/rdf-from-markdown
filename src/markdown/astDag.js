import rdf from 'rdf-ext'
import ns from '../namespaces.js'
import { getList, getText } from './ast.js'
import {getTags} from '../text/tags.js'

async function getAstDag ({ astNode, fullText }) {

  const root = rdf.blankNode()
  const dataset = rdf.dataset()
  const pointer = rdf.clownface({ term: root, dataset })
  const quads = []
  quads.push(rdf.quad(root, ns.ex.label, rdf.literal('ROOT')))

  let headersStack = []
  astNode.children.reduce((currentUri, astNode) => {
    const text = getText({ astNode, fullText })

    if (astNode.type === 'heading') {
      const currentIri = rdf.blankNode()

      const ancesters = headersStack.filter(x => x.depth < astNode.depth)
      if (ancesters.length) {
        headersStack = headersStack.filter(x => x.depth < astNode.depth)
      }
      const parentIri = ancesters.length
        ? ancesters[ancesters.length - 1].iri
        : currentUri

      quads.push(rdf.quad(parentIri, ns.mark.contains, currentIri))
      quads.push(rdf.quad(currentIri, ns.mark.header, rdf.literal(text)))

      for (const tag of getTags(text)){
        quads.push(rdf.quad(currentIri, ns.mark.tag, rdf.literal(tag)))
      }

      headersStack.push({
        iri: currentIri,
        depth: astNode.depth,
      })
      return currentIri
    } else {
      if (astNode.type === 'list') {
        for (const text of getList({ astNode: astNode, fullText: fullText })) {
          quads.push(rdf.quad(currentUri, ns.mark.text, rdf.literal(text)))
        }
      } else if (astNode.type === 'paragraph') {
        quads.push(rdf.quad(currentUri, ns.mark.text, rdf.literal(text)))
      }
    }
    return currentUri
  }, root)
  pointer.dataset.addAll(quads)
  return pointer
}

export { getAstDag }
