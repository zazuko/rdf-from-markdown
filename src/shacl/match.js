import rdf from 'rdf-ext'
import ns from '../namespaces.js'
import { extractInlineFields } from '../text/inlineFields.js'
import { normalizeText } from '../text/normalize.js'

function isNormalizedMatch (first, second) {
  return normalizeLiteral(first).value.toLowerCase() ===
    normalizeLiteral(second).value.toLowerCase()
}

function normalizeLiteral (term) {
  return rdf.literal(normalizeText(term.value))
}

function produceQuads ({ uri, astPointer, shapePointer, uriResolver }) {

  const quads = []
  const entityType = shapePointer.out(ns.sh.class).term
  quads.push(rdf.quad(uri, ns.rdf.type, entityType))

  for (const property of shapePointer.out(ns.sh.property).terms) {
    const propertyCF = shapePointer.node(property)
    quads.push(...matchProperties(
      { uri, uriResolver, astPointer, shapePointer: propertyCF }))
  }
  return quads
}

const defaultMatchers = [
  {
    matchProperty: ns.mark.attachToHeader, matchTarget: ns.mark.header,
  }, {
    matchProperty: ns.mark.attachToItem, matchTarget: ns.mark.item,
  }, {
    matchProperty: ns.mark.attachToListItem, matchTarget: ns.mark.listItem,
  }]

function matchProperties ({
  uri, uriResolver, astPointer, shapePointer,
}) {
  const quads = []

  for (const { matchProperty, matchTarget } of defaultMatchers) {
    const matchHeader = shapePointer.out(matchProperty).term
    if (matchHeader) {
      const matchesAll = matchHeader.equals(ns.mark.all)
      const astHeader = astPointer.out(ns.mark.header).term

      if (!matchesAll && astHeader) {
        quads.push(rdf.quad(uri, ns.schema.name, normalizeLiteral(astHeader)))
      }

      for (const childUri of astPointer.out(ns.mark.contains).terms) {
        const childAstPointer = astPointer.node(childUri)
        if (matchesAll ||
          isNormalizedMatch(childAstPointer.out(matchTarget).term,
            matchHeader)) {
          for (const included of shapePointer.out(ns.mark.include).terms) {
            for (const text of childAstPointer.out(included).terms) {
              quads.push(...getValue({
                uri,
                shapePointer,
                uriResolver,
                text,
                astPointer: childAstPointer,
              }))
            }
          }
        }
      }

    }
  }
  return quads
}

function getValue ({
  uri, uriResolver, astPointer, shapePointer, text,
}) {
  const path = shapePointer.out(ns.sh.path).term
  const quads = []
  const nodeKind = shapePointer.out(ns.sh.nodeKind).term
  const normalizedLiteral = normalizeLiteral(text)

  if (nodeKind.equals(ns.sh.Literal)) {
    quads.push(rdf.quad(uri, path, normalizedLiteral))
  } else if (nodeKind.equals(ns.sh.IRI)) {
    const childUri = uriResolver.mintUri(normalizedLiteral)
    quads.push(rdf.quad(uri, path, childUri))
    quads.push(rdf.quad(childUri, ns.schema.name, normalizedLiteral))
    const inlineQuads = getQuadsFromInline(
      { uri: childUri, literal: text, shapes: shapePointer.any() })
    quads.push(...inlineQuads)

    const node = shapePointer.out(ns.sh.node).term
    if (node) {
      quads.push(...produceQuads({
        uri: childUri,
        astPointer,
        shapePointer: shapePointer.node(node),
        uriResolver,
      }))
    }
  }
  return quads
}

function getQuadsFromInline ({ uri, literal, shapes }) {

  const result = []
  for (const inlineField of extractInlineFields(literal.value)) {
    const normalizedProperty = rdf.literal(normalizeText(inlineField.property))
    for (const current of shapes.dataset.match(null,
      ns.mark.matchInlineProperty, normalizedProperty)) {
      const normalizedValue = rdf.literal(normalizeText(inlineField.value))
      const path = shapes.node(current.subject).out(ns.sh.path).term
      result.push(rdf.quad(uri,path,normalizedValue))
    }
  }
  return result
}

function getShapesByProperty ({ property, value, shapes }) {
  const matches = []
  for (const current of shapes.dataset.match(null, property, value)) {
    matches.push(shapes.node(current.subject))
  }
  return matches
}

function getShapesByTag ({ tag, shapes }) {
  return getShapesByProperty(
    { property: ns.mark.matchHashTag, value: tag, shapes })
}

function getShapesByType ({ type, shapes }) {
  return getShapesByProperty(
    { property: ns.mark.matchType, value: type, shapes })
}

const defaultUriResolver = {
  mintUri (literal) {
    return rdf.blankNode()
  },
}

function doShaclMatch ({
  astPointer, shapes, uriResolver = defaultUriResolver,
}) {
  const result = []

  for (const type of astPointer.out(ns.rdf.type).terms) {
    for (const shapeRoot of getShapesByType({ type, shapes })) {
      const uri = uriResolver.mintUri(type)
      const quads = produceQuads(
        { uri, astPointer, shapePointer: shapeRoot, uriResolver })
      result.push(...quads)
    }
  }

  // Check if a tag is found
  for (const tag of astPointer.out(ns.mark.tag).terms) {
    for (const shapeRoot of getShapesByTag({ tag, shapes })) {
      const uri = uriResolver.mintUri(tag)
      const quads = produceQuads(
        { uri, astPointer, shapePointer: shapeRoot, uriResolver })
      result.push(...quads)
    }
  }

  // Recursively process the tree
  for (const term of astPointer.out(ns.mark.contains).terms) {
    const quads = doShaclMatch(
      { astPointer: astPointer.node(term), shapes, uriResolver })
    result.push(...quads)
  }

  return result
}

export { doShaclMatch }
