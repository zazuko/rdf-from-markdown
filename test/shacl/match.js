import { expect } from 'expect'

import toMatchSnapshot from 'expect-mocha-snapshot'
import { resolve } from 'path'
import rdf from 'rdf-ext'
import { getAstDag } from '../../src/markdown/astDag.js'
import { createMarkdownParser } from '../../src/markdown/markdownParser.js'
import { doShaclMatch } from '../../src/shacl/match.js'
import { getClownface, getText } from '../support/readFiles.js'
import { prettyPrint } from '../util.js'
import {tests} from '../support/tests.js'

expect.extend({ toMatchSnapshot })

describe('match', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const fullText = await getText(
        { path: resolve(current.markdown) })
      const shapesClownface = await getClownface(
        { path: resolve(current.shacl) })
      const parser = createMarkdownParser()
      const ast = await parser.parse(fullText)

      const astGraph = await getAstDag({ astNode: ast, fullText })
      const quads = doShaclMatch({ astPointer:astGraph, shapes:shapesClownface})

      const dataset = rdf.dataset().addAll(quads)
      const turtle = await prettyPrint(dataset)
      expect(turtle).toMatchSnapshot(this)
    })
  }
})
