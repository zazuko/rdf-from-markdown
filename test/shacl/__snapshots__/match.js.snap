// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match can use a taxonomy (recursive patterns) 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/named/1> a ex:Thing ;
	ex:hasChildren <http://example.org/named/2> .

<http://example.org/named/2> a ex:Thing ;
	ex:hasChildren <http://example.org/named/3>, <http://example.org/named/8> ;
	schema:name \\"Things\\" .

<http://example.org/named/3> a ex:Thing ;
	ex:hasChildren <http://example.org/named/4>, <http://example.org/named/6>, <http://example.org/named/7> ;
	schema:name \\"Food\\" .

<http://example.org/named/4> a ex:Thing ;
	ex:hasChildren <http://example.org/named/5> ;
	schema:name \\"Red\\" .

<http://example.org/named/5> a ex:Thing ;
	schema:name \\"Tomato\\" .

<http://example.org/named/6> a ex:Thing ;
	schema:name \\"Green\\" .

<http://example.org/named/7> a ex:Thing ;
	schema:name \\"Lettuce\\" .

<http://example.org/named/8> a ex:Thing ;
	ex:hasChildren <http://example.org/named/9>, <http://example.org/named/10>, <http://example.org/named/11>, <http://example.org/named/12>, <http://example.org/named/13>, <http://example.org/named/14>, <http://example.org/named/15> ;
	schema:name \\"Tools\\" .

<http://example.org/named/9> a ex:Thing ;
	schema:name \\"Metal\\" .

<http://example.org/named/10> a ex:Thing ;
	schema:name \\"* Screwdriver\\" .

<http://example.org/named/11> a ex:Thing ;
	schema:name \\"* Hammer\\" .

<http://example.org/named/12> a ex:Thing ;
	schema:name \\"Wood\\" .

<http://example.org/named/13> a ex:Thing ;
	schema:name \\"1\\" .

<http://example.org/named/14> a ex:Thing ;
	schema:name \\"2\\" .

<http://example.org/named/15> a ex:Thing ;
	schema:name \\"3\\" .

"
`;

exports[`match produces quads using embedded tag 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/named/1> a ex:Role ;
	schema:name \\"Pancake maker\\" ;
	ex:assignment <http://example.org/named/2>, <http://example.org/named/3> ;
	ex:purpose \\"To make the best Pancakes in the world\\" ;
	ex:tasks \\"And a paragraph, explaining even more things\\", \\"Mix ingredients\\", \\"Heat the pan\\", \\"Prepare the Pancakes\\" .

<http://example.org/named/2> schema:name \\"Bob\\" ;
	ex:action \\"prepare\\" .

<http://example.org/named/3> schema:name \\"Alice\\" ;
	ex:action \\"cook\\" .

<http://example.org/named/4> a ex:Role ;
	schema:name \\"Pancake seller\\" ;
	ex:assignment <http://example.org/named/5> ;
	ex:purpose \\"Sell the Pancakes to the customers\\" ;
	ex:tasks \\"Coordination of Pancake economic activities\\", \\"Advertise the Pancakes\\" .

<http://example.org/named/5> schema:name \\"Charlie\\" .

"
`;

exports[`match produces quads using lists 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/named/1> a ex:People ;
	schema:hasPart <http://example.org/named/2>, <http://example.org/named/4>, <http://example.org/named/6> .

<http://example.org/named/2> a ex:Person ;
	schema:name \\"Alice\\" ;
	ex:content <http://example.org/named/3> .

<http://example.org/named/3> schema:name \\"\\" ;
	ex:hasAge \\"22\\" .

<http://example.org/named/4> a ex:Person ;
	schema:name \\"Bob\\" ;
	ex:content <http://example.org/named/5> ;
	ex:hasAge \\"42\\" .

<http://example.org/named/5> schema:name \\"Eating\\" .

<http://example.org/named/6> a ex:Person ;
	schema:name \\"Charlie\\" ;
	ex:content <http://example.org/named/7> .

<http://example.org/named/7> schema:name \\"Charlie is a good guy.\\" ;
	ex:hasAge \\"56\\" .

"
`;
